File: train.py
#!/usr/bin/env python

import time
from pathlib import Path

import typer

from quicktok import BasicTokenizer


def main(
    data_file_path: Path,
    model_file_path: Path = Path("models/taylorswift.model"),
    num_threads: int = 1,
    vocab_size: int = 512,
    save_vocab: bool = False,
):
    data_dir_path = Path("data")
    data_file_path = data_dir_path / "taylorswift.txt"

    text = data_file_path.read_text(encoding="utf-8")
    model_file_path.parent.mkdir(parents=True, exist_ok=True)
    vocab_file_path = str(model_file_path.with_suffix(".vocab")) if save_vocab else None
    t0 = time.time()
    tokenizer = BasicTokenizer(num_threads)
    tokenizer.train(text, vocab_size)
    tokenizer.save(str(model_file_path), vocab_file_path)
    t1 = time.time()
    print(f"Training took {t1 - t0:.2f} seconds")


if __name__ == "__main__":
    typer.run(main)

File: examples/src/bin/train.rs
use std::{fs, path::PathBuf, time::Instant};

use clap::Parser;
use quicktok::{BasicTokenizer, Tokenizer};

#[derive(Parser, Debug)]
#[clap(author, version, about, long_about = None)]
struct Args {
    data_file_path: PathBuf,

    model_file_path: PathBuf,

    #[clap(long, default_value_t = 1)]
    num_threads: usize,

    #[clap(long, default_value_t = 512)]
    vocab_size: usize,

    #[clap(long)]
    save_vocab: bool,
}

fn main() {
    let args = Args::parse();
    let text = fs::read_to_string(&args.data_file_path).expect("failed to read data file");
    let output_dir_path = args
        .model_file_path
        .parent()
        .expect("failed to get output directory");
    fs::create_dir_all(output_dir_path).expect("failed to create output directory");
    let vocab_file_path = args.model_file_path.with_extension("vocab");
    let start_time = Instant::now();
    let mut tokenizer = BasicTokenizer::new(args.num_threads);
    tokenizer.train(&text, args.vocab_size);
    tokenizer
        .save(
            args.model_file_path.to_str().unwrap(),
            if args.save_vocab {
                Some(vocab_file_path.to_str().unwrap())
            } else {
                None
            },
        )
        .expect("failed to save tokenizer model and vocabulary");
    let duration = start_time.elapsed();
    println!("Training took {:.2} seconds", duration.as_secs_f32());
}
